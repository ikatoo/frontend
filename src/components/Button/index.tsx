import { ButtonHTMLAttributes } from 'react'
import Styles from './styles'

type ButtonProps = {
  children?: React.ReactNode
  styleType?: 'primary' | 'secondary' | 'default'
  icon?: JSX.Element
  block?: boolean
  tabIndex?: number
  asLink?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

const Button = ({
  children,
  icon,
  styleType = 'default',
  block = false,
  tabIndex,
  ...props
}: ButtonProps) => {
  return (
    <Styles.Wrapper
      {...props}
      tabIndex={tabIndex}
      $block={block}
      $styleType={styleType}
    >
      {!!icon && <Styles.IconWrapper>{icon}</Styles.IconWrapper>}
      {children}
    </Styles.Wrapper>
  )
}

export default Button
