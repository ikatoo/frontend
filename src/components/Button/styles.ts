import tw from 'tailwind-styled-components'

const type = {
  primary: `
    text-white
    bg-blue-800
    hover:bg-mck_aqua
    hover:text-slate-800
    focus:bg-blue-700
    active:bg-blue-800
  `,
  secondary: `
    text-mck_aqua
    bg-slate-800
    hover:bg-mck_aqua
    hover:text-slate-800
    focus:bg-blue-700
    active:bg-blue-800
  `,
  default: `
    text-black
    bg-gray-500
  `
}

export default {
  Wrapper: tw.button`
    flex
    flex-row
    justify-center
    items-center
    gap-4
    px-7
    py-3
    ${(props: { $styleType: 'primary' | 'secondary' | 'default' }) =>
      type[props.$styleType]}
    font-medium
    text-sm
    leading-snug
    uppercase
    rounded
    shadow-mck_black
    shadow
    hover:shadow-md
    focus:shadow-md
    focus:ring-0
    active:shadow-md
    transition
    duration-150
    ease-in-out
    ${(props: { $block: boolean }) => !!props.$block && 'w-full'}
    disabled:bg-mck_black
    disabled:text-mck_gray_dark
  `,
  IconWrapper: tw.div`
    flex
    flex-col
    w-5
    h-full
  `
}
