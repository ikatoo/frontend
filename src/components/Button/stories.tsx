import { Github } from '@styled-icons/boxicons-logos'
import { Angry, Laugh } from '@styled-icons/boxicons-regular'
import Button from '.'

import type { Meta, StoryObj } from '@storybook/react'

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button
}

const Wrapper = (props: { children: React.ReactNode }) => (
  <div className="flex justify-center items-center bg-mck_black_light h-screen w-full">
    <div className="flex flex-col w-fit h-fit gap-2">{props.children}</div>
  </div>
)

export default meta
type Story = StoryObj<typeof Button>

export const Default: Story = {
  render: () => <Button />
}

export const With_Label: Story = {
  render: () => <Button>Button</Button>
}

export const With_Icon: Story = {
  render: () => (
    <div>
      <Button icon={<Github />}>GitHub</Button>
    </div>
  )
}

export const Block: Story = {
  render: () => (
    <div>
      <Button block>GitHub</Button>
    </div>
  )
}

export const Types: Story = {
  render: () => (
    <Wrapper>
      <Button styleType="default">Default</Button>
      <Button styleType="primary" icon={<Laugh />}>
        Primary
      </Button>
      <Button styleType="secondary" icon={<Angry />}>
        Secondary
      </Button>
    </Wrapper>
  )
}

export const Disabled: Story = {
  render: () => (
    <Wrapper>
      <Button disabled styleType="default">
        Default
      </Button>
      <Button disabled styleType="primary" icon={<Laugh />}>
        Primary
      </Button>
      <Button disabled styleType="secondary" icon={<Angry />}>
        Secondary
      </Button>
    </Wrapper>
  )
}
