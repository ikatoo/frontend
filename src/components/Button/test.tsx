import { render, screen } from '@testing-library/react'

import Button from '.'
import { Github } from '@styled-icons/boxicons-logos'

describe('<Button />', () => {
  it('should render the component', () => {
    render(<Button />)
    const button = screen.getByRole('button')

    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('text-black')
    expect(button).toHaveClass('bg-gray-500')
  })

  it('should render primary button', () => {
    render(<Button styleType="primary" />)
    const button = screen.getByRole('button')

    expect(button).toHaveClass('text-mck_aqua')
    expect(button).toHaveClass('bg-slate-800')
    expect(button).toHaveClass('hover:bg-mck_aqua')
    expect(button).toHaveClass('hover:text-slate-800')
    expect(button).toHaveClass('focus:bg-blue-700')
    expect(button).toHaveClass('active:bg-blue-800')
  })

  it('should render secondary button', () => {
    render(<Button styleType="secondary" />)
    const button = screen.getByRole('button')

    expect(button).toHaveClass('text-white')
    expect(button).toHaveClass('bg-blue-800')
    expect(button).toHaveClass('hover:bg-mck_aqua')
    expect(button).toHaveClass('hover:text-slate-800')
    expect(button).toHaveClass('focus:bg-blue-700')
    expect(button).toHaveClass('active:bg-blue-800')
  })

  it('should render button with icon', () => {
    render(<Button icon={<Github data-testid="icon" />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render block button', () => {
    render(<Button block />)

    expect(screen.getByRole('button')).toHaveClass('w-full')
  })

  it('should render button and match with snapshot', () => {
    const { container } = render(<Button block />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
