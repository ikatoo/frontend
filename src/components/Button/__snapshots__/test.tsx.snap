// Vitest Snapshot v1

exports[`<Button /> > should render button and match with snapshot 1`] = `
<button
  class="flex flex-row justify-center items-center gap-4 px-7 py-3 text-black bg-gray-500 focus:outline-none font-medium text-sm leading-snug uppercase rounded shadow-md hover:shadow-lg focus:shadow-lg focus:ring-0 active:shadow-lg transition duration-150 ease-in-out w-full"
/>
`
